app/services/action_collator.rb

```diff
@@ -0,0 +1,47 @@
+# frozen_string_literal: true
+class ActionCollator
+  PRIVATE_KEYS = %w(action_referrer_email action_express_donation).freeze
+  PREFIXES = %w(action_textentry_ action_box_ action_dropdown_ action_choice_ action_).freeze
+
+  def self.run(actions)
+    new(actions).run
+  end
+
+  def self.csv(actions)
+    new(actions).csv
+  end
+
+  def initialize(actions)
+    @actions = actions
+  end
+
+  def run
+    return hashes, keys, headers
+  end
+
+  def csv
+    rows = @actions.map do |action|
+      keys.map { |k| action.form_data[k] }
+    end
+    rows.unshift(headers).map { |r| r.join(',') }.join("\n")
+  end
+
+  def hashes
+    @hashes ||= @actions.map do |action|
```

Comment:
In order to avoid loading all action AR objects in memory at the same time, we could use the ActiveRecord `in_batches` and `pluck` methods. 
```
@actions.in_batches.pluck(:form_data) do |action|
```
http://api.rubyonrails.org/classes/ActiveRecord/Batches.html#method-i-in_batches